### PR TITLE
HPCC-20269 Initial ESP XSD review

### PR DESCRIPTION
**Pull request by kenrowland**
Add initial ESP XSD including necessary supporting changes

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by kenrowland** [comment]
@rpastrana @RussWhitehead @JamesDeFabia @wangkx 

Please review. See accompanying email for more information

**Comment by kenrowland** [comment]
@afishbeck Please review as well. You are probably most interested in the esp.xsd.

**Comment by kenrowland** [comment]
@rpastrana Updating last person.
Per our discussion, this review is only for the conversion of the XSD. Another JIRA will be opened for Keven and/or Tony to review its correctness.

**Comment by rpastrana** [comment]
@kenrowland please rebase so I can review

**Review by rpastrana — COMMENTED** [review]
@kenrowland a few comments

File: configuration/config2/ConfigPath.cpp [hunk 1]
```diff
@@ -20,20 +20,22 @@
...
-    for (auto it=m_attributeValues.begin(); it!=m_attributeValues.end(); ++it)
+    for (auto it=m_attributeValues.begin(); it!=m_attributeValues.end() && !found; ++it)
     {
-        if (*it == val)
-            return true;
+        found = (*it == val);
     }
 
-    return false;
+    bool rc = found == m_presentInList;
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/config2/ConfigPath.cpp]: It is, and some future work has changed this method. For now I will leave this as is. Future PR will update this method

File: configuration/config2/ConfigPath.cpp [hunk 2]
```diff
@@ -118,10 +120,12 @@ void ConfigPath::parsePathElement(std::size_t start, const std::shared_ptr<Confi
 
             //
             // Value present?
+            std::size_t notPos = attr.find_first_of('!');
             std::size_t equalPos = attr.find_first_of('=');
             if (equalPos != std::string::npos)
             {
-                pPathItem->setAttributeName(attr.substr(1, equalPos-1));
+                std::size_t compareLength = (notPos == std::string::npos) ? 1 : 2;
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/config2/ConfigPath.cpp]: I don't think so, unless we want to do validate all XPaths as being well formed.

File: configuration/config2/XSDSchemaParser.cpp [hunk 3]
```diff
@@ -246,7 +247,29 @@ void XSDSchemaParser::parseAttributeGroup(const pt::ptree &attributeTree)
...
+                        std::vector<std::string> vals = splitString(info, "=");
+                        if (vals.size() != 2)
+                        {
+                            throw (ParseException("Invalid default value override in attribute group (" + refName + ")"));
+                        }
+
+                        std::shared_ptr<SchemaValue> pAttr = m_pSchemaItem->getAttribute(vals[0], false);
+                        if (pAttr)
+                        {
+                            pAttr->setCodeDefault(vals[1]);
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/config2/XSDSchemaParser.cpp]: That change has not been implemented in this PR as it was decided after this PR was made. It is coming in a later PR that will address all current XSDs and the code necessary.

File: initfiles/componentfiles/configschema/xsd/esp.xsd [hunk 4]
```diff
@@ -1,18 +1,33 @@
...
         <xs:sequence>
             <xs:element name="EspProcess" hpcc:class="component" hpcc:category="ESP Process" hpcc:itemType="esp" hpcc:displayName="ESP Process"  hpcc:docid="ESP.t6" maxOccurs="unbounded">
-                <xs:annotation>
-                    <xs:documentation>ESP Service Bindings</xs:documentation>
-                </xs:annotation>
                 <xs:complexType>
                     <xs:sequence>
 
-                        <xs:element name="EspBinding" maxOccurs="unbounded" hpcc:itemType="espbinding" hpcc:category="ESP Bindings" hpcc:insertLimitType="attribute" hpcc:insertLimitData="service" hpcc:displayName="ESP Service Bindings" hpcc:class="elementSet" hpcc:event="create">
+                        <xs:element name="EspBinding" maxOccurs="unbounded" hpcc:itemType="espbinding" hpcc:category="ESP Bindings" hpcc:insertLimitType="attribute" hpcc:insertLimitData="service"
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply initfiles/componentfiles/configschema/xsd/esp.xsd]: OK, just make sure that a Config Manager 2.0 Jira is opened so it is tracked.

**Review by kenrowland — COMMENTED** [review]
@rpastrana Please see replies.

**Comment by kenrowland** [comment]
@rpastrana Rebased. Once merged, the next PR can also be rebased into a single commit also.

File: configuration/config2/ConfigPath.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rpastrana) [reply configuration/config2/ConfigPath.cpp]: go ahead and clean it up now.

File: configuration/config2/ConfigPath.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rpastrana) [reply configuration/config2/ConfigPath.cpp]: is it possible attr can contain '=!' or even '!xyz='? If so, make sure you're not making any assumptions about notpos and equalpos

File: initfiles/componentfiles/configschema/xsd/esp.xsd (hunk 4 above)
[the remark this answers is not in the document]
Comment (rpastrana) [reply initfiles/componentfiles/configschema/xsd/esp.xsd]: you have to make sure the version of config2.0 XSDs reflect the content of the configmgr XSDs at the time your release config 2.0

**Review by rpastrana — CHANGES_REQUESTED** [review]
@kenrowland

**Comment by kenrowland** [comment]
@rpastrana changed return code and processing of attribute values list. Please review. All other open comments are noted.

**Review by rpastrana — APPROVED** [review]
@kenrowland

**Comment by kenrowland** [comment]
@richardkchapman please merge

**Pull request merged**